### PR TITLE
add correlation function

### PR DESCRIPTION
**Pull request by realno**
# Which issue does this PR close?

This PR add the `corr` function which is part of https://github.com/apache/arrow-datafusion/issues/1486

 # Rationale for this change

Add correlation function to DataFusion.

# What changes are included in this PR?

New function `corr(col1, co2)` is added.

# Are there any user-facing changes?

New function `corr(col1, co2)` is added.

No breaking change.

### CODE REVIEW

**Review by houqp — APPROVED** [review]

**Comment by liukun4515** [comment]
why not support `covar_pop` and `covar_samp` togethe?
reference https://www.postgresql.org/docs/14/functions-aggregate.html
@realno

**Comment by realno** [comment]
> why not support `covar_pop` and `covar_samp` together? reference https://www.postgresql.org/docs/14/functions-aggregate.html @realno

`covar_pop` and `covar_samp` were merged here https://github.com/apache/arrow-datafusion/pull/1551

**Review by liukun4515 — APPROVED** [review]
LGTM
you can rebase the master, after the Clippy fixed

File: datafusion/src/physical_plan/expressions/correlation.rs [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+
+    fn evaluate(&self) -> Result<ScalarValue> {
+        let covar = self.covar.evaluate()?;
+        let stddev1 = self.stddev1.evaluate()?;
+        let stddev2 = self.stddev2.evaluate()?;
+
+        if let ScalarValue::Float64(Some(c)) = covar {
+            if let ScalarValue::Float64(Some(s1)) = stddev1 {
+                if let ScalarValue::Float64(Some(s2)) = stddev2 {
+                    return Ok(ScalarValue::Float64(Some(c / s1 / s2)));
```
Comment (Jimexist) [inline datafusion/src/physical_plan/expressions/correlation.rs:223]: can s1 or s2 be 0?

Comment (realno) [reply datafusion/src/physical_plan/expressions/correlation.rs]: They may be 0, pushed a fix. Thanks!

**Comment by alamb** [comment]
Thanks @realno !

**Pull request merged**